experimental/single_number_formatter/src/ule.rs

```diff
@@ -0,0 +1,130 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use zerovec::{
+    maps::ZeroMapKV,
+    ule::{AsULE, ZeroVecError, ULE},
+};
+
+use crate::provider::{CurrencyPatterns, PatternSelection};
+
+/// `CurrencyPatternsULE` is a type optimized for efficient storing and
+/// deserialization of `CurrencyPatterns` using the `ZeroVec` model.
+///
+/// The serialization model packages the pattern item in three bytes.
+///
+/// The first two bits is used to determine the short_pattern_standard. If the bits are `00`, then, the value will be `Standard`.
+/// If the bits are `01`, then, the value will be `StandardAlphaNextToNumber`.
+///
+/// The second two bits is used to determine the narrow_pattern_standard. If the bits are `00`, then, the value will be `Standard`.
+/// If the bits are `01`, then, the value will be `StandardAlphaNextToNumber`.
+///
+/// The third two bits with the second byte is used to determine the short_place_holder_index.
+/// The fourth two bits with the third byte is used to determine the narrow_place_holder_index.
+#[derive(Copy, Clone, Debug, PartialEq)]
+#[repr(transparent)]
+pub struct CurrencyPatternsULE([u8; 3]);
+
+impl CurrencyPatternsULE {}
+
+// Safety (based on the safety checklist on the ULE trait):
+//  1. CurrencyPatternsULE does not include any uninitialized or padding bytes.
+//     (achieved by `#[repr(transparent)]` on a ULE type)
+//  2. CurrencyPatternsULE is aligned to 1 byte.
+//     (achieved by `#[repr(transparent)]` on a ULE type)
+//  3. The impl of validate_byte_slice() returns an error if any byte is not valid.
+//  4. The impl of validate_byte_slice() returns an error if there are extra bytes.
+//  5. The other ULE methods use the default impl.
+//  6. CurrencyPatternsULE byte equality is semantic equality.
+unsafe impl ULE for CurrencyPatternsULE {
+    fn validate_byte_slice(bytes: &[u8]) -> Result<(), ZeroVecError> {
+        if bytes.len() % 3 != 0 {
+            return Err(ZeroVecError::length::<Self>(bytes.len()));
+        }
+
+        Ok(())
+    }
+}
+
+const PATTERN_SHORT_POSITION: u8 = 6;
+const PATTERN_NARROW_POSITION: u8 = 4;
+const INDEX_SHORT_POSITION: u8 = 2;
+const INDEX_NARROW_POSITION: u8 = 0;
+
+impl AsULE for CurrencyPatterns {
+    type ULE = CurrencyPatternsULE;
+
+    #[inline]
+    fn to_unaligned(self) -> Self::ULE {
+        let mut first_byte_ule: u8 = 0;
+
+        if self.short_pattern_standard == PatternSelection::StandardAlphaNextToNumber {
+            first_byte_ule |= 0b01 << PATTERN_SHORT_POSITION;
```

Comment:
```suggestion
            first_byte_ule |= 1 << PATTERN_SHORT_POSITION;
```

Reply:
I prefer the binary form.

Reply:
I'd find this more readable with `0b0100_0000` instead of bit shifting by a constant. However, what you have here is fine, and I think using `0b01` is good because it emphasizes that we are putting 2 bits in place.

Reply:
I think `0b0100_0000` is more readable, but the I want to unify the shifting with constants for the future maintenance.